panic: runtime error: makeslice: len out of range

Panic found with #101 :+1: 
```
panic: runtime error: makeslice: len out of range

goroutine 1 [running]:
github.com/bio-routing/bio-rd/protocols/bgp/packet.deserializeMultiProtocolUnreachNLRI(0xc420018540, 0x1, 0x1, 0x1, 0x1, 0x0, 0x0, 0x5b6bb0, 0x4d8f20)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/mp_unreach_nlri.go:33 +0x8c
github.com/bio-routing/bio-rd/protocols/bgp/packet.(*PathAttribute).decodeMultiProtocolUnreachNLRI(0xc42000c180, 0xc4200b0150, 0x1, 0x0)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/path_attributes.go:165 +0x1c2
github.com/bio-routing/bio-rd/protocols/bgp/packet.decodePathAttr(0xc4200b0150, 0xc42005bee5, 0x5181e0, 0x5b6bb0, 0x4d8e60, 0xc4200885b0)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/path_attributes.go:116 +0xa59
github.com/bio-routing/bio-rd/protocols/bgp/packet.decodePathAttrs(0xc4200b0150, 0xc420053030, 0xc42005bee5, 0x1, 0x0, 0x0)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/path_attributes.go:23 +0x6d
github.com/bio-routing/bio-rd/protocols/bgp/packet.decodeUpdateMsg(0xc4200b0150, 0x1d, 0xc42005bee5, 0x68, 0x12, 0x4d8e60)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/decoder.go:63 +0x15f
github.com/bio-routing/bio-rd/protocols/bgp/packet.decodeMsgBody(0xc4200b0150, 0xc4001d8402, 0xc42005bee5, 0x0, 0x5, 0x6, 0xc420014480)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/decoder.go:36 +0xe7
github.com/bio-routing/bio-rd/protocols/bgp/packet.Decode(0xc4200b0150, 0xc42005bee5, 0xa, 0x45f576, 0x5b52081b)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/decoder.go:20 +0x102
github.com/bio-routing/bio-rd/protocols/bgp/packet.Fuzz(0x7f17bc462000, 0x1b, 0x200000, 0x3)
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/fuzzing.go:22 +0xbd
go-fuzz-dep.Main(0x50ac10)
	/tmp/go-fuzz-build543490251/goroot/src/go-fuzz-dep/main.go:49 +0xad
main.main()
	/tmp/go-fuzz-build543490251/gopath/src/github.com/bio-routing/bio-rd/protocols/bgp/packet/go.fuzz.main/main.go:10 +0x2d
exit status 2```
Quoted input:
```
	"\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x000\x02\x00" +
	"\x0000\x00\x0f\x010"
```

--- a/protocols/bgp/packet/mp_reach_nlri.go
+++ b/protocols/bgp/packet/mp_reach_nlri.go
@@ -38,8 +38,13 @@ func (n *MultiProtocolReachNLRI) serialize(buf *bytes.Buffer) uint16 {
 func deserializeMultiProtocolReachNLRI(b []byte) (MultiProtocolReachNLRI, error) {
 	n := MultiProtocolReachNLRI{}
 	nextHopLength := uint8(0)
-	variable := make([]byte, len(b)-4)
 
+	variableLength := len(b) - 4 // 4 <- AFI + SAFI + NextHopLength
+	if variableLength <= 0 {
+		return n, fmt.Errorf("Invalid length of MP_REACH_NLRI: expected more than 4 bytes but got %d", len(b))
+	}
+
+	variable := make([]byte, variableLength)
 	fields := []interface{}{
 		&n.AFI,
 		&n.SAFI,
@@ -51,18 +56,24 @@ func deserializeMultiProtocolReachNLRI(b []byte) (MultiProtocolReachNLRI, error)
 		return MultiProtocolReachNLRI{}, err
 	}
 
+	budget := variableLength
+	if budget < int(nextHopLength) {
+		return MultiProtocolReachNLRI{},
+			fmt.Errorf("Failed to decode next hop IP: expected %d bytes for NLRI, only %d remaining", nextHopLength, budget)
+	}
+
 	n.NextHop, err = bnet.IPFromBytes(variable[:nextHopLength])
 	if err != nil {
 		return MultiProtocolReachNLRI{}, fmt.Errorf("Failed to decode next hop IP: %v", err)
 	}
-
-	variable = variable[1+nextHopLength:]
+	budget -= int(nextHopLength)
 
 	n.Prefixes = make([]bnet.Prefix, 0)
-
-	if len(variable) == 0 {
+	if budget == 0 {
 		return n, nil
 	}
+
+	variable = variable[1+nextHopLength:] // 1 <- RESERVED field
 
 	idx := uint16(0)
 	for idx < uint16(len(variable)) {

--- a/protocols/bgp/packet/mp_unreach_nlri.go
+++ b/protocols/bgp/packet/mp_unreach_nlri.go
@@ -30,36 +30,41 @@ func (n *MultiProtocolUnreachNLRI) serialize(buf *bytes.Buffer) uint16 {
 
 func deserializeMultiProtocolUnreachNLRI(b []byte) (MultiProtocolUnreachNLRI, error) {
 	n := MultiProtocolUnreachNLRI{}
-	prefix := make([]byte, len(b)-3)
 
+	prefixesLength := len(b) - 3 // 3 <- AFI + SAFI
+	if prefixesLength < 0 {
+		return n, fmt.Errorf("Invalid length of MP_UNREACH_NLRI: expected more than 3 bytes but got %d", len(b))
+	}
+
+	prefixes := make([]byte, prefixesLength)
 	fields := []interface{}{
 		&n.AFI,
 		&n.SAFI,
-		&prefix,
+		&prefixes,
 	}
 	err := decode(bytes.NewBuffer(b), fields)
 	if err != nil {
 		return MultiProtocolUnreachNLRI{}, err
 	}
 
-	if len(prefix) == 0 {
+	if len(prefixes) == 0 {
 		return n, nil
 	}
 
 	idx := uint16(0)
-	for idx < uint16(len(prefix)) {
-		pfxLen := prefix[idx]
+	for idx < uint16(len(prefixes)) {
+		pfxLen := prefixes[idx]
 		numBytes := uint16(BytesInAddr(pfxLen))
 		idx++
 
-		r := uint16(len(prefix)) - idx
+		r := uint16(len(prefixes)) - idx
 		if r < numBytes {
 			return MultiProtocolUnreachNLRI{}, fmt.Errorf("expected %d bytes for NLRI, only %d remaining", numBytes, r)
 		}
 
 		start := idx
 		end := idx + numBytes
-		pfx, err := deserializePrefix(prefix[start:end], pfxLen, n.AFI)
+		pfx, err := deserializePrefix(prefixes[start:end], pfxLen, n.AFI)
 		if err != nil {
 			return MultiProtocolUnreachNLRI{}, err
 		}

--- a/protocols/bgp/packet/path_attributes_test.go
+++ b/protocols/bgp/packet/path_attributes_test.go
@@ -968,6 +968,17 @@ func TestDecodeMultiProtocolReachNLRI(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "MP_REACH_NLRI with invalid prefixes",
+			input: []byte{
+				0x00, 0x02, // AFI
+				0x01,                                                                                                 // SAFI
+				0x10, 0x20, 0x01, 0x06, 0x78, 0x01, 0xe0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, // NextHop
+				0x00,             // RESERVED
+				0x30, 0x26, 0x00, // Prefix
+			},
+			wantFail: true,
+		},
 	}
 
 	t.Parallel()
@@ -1036,6 +1047,15 @@ func TestDecodeMultiProtocolUnreachNLRI(t *testing.T) {
 			name: "MP_UNREACH_NLRI with invalid length",
 			input: []byte{
 				0x00, 0x02, // AFI
+			},
+			wantFail: true,
+		},
+		{
+			name: "MP_REACH_NLRI with invalid prefixes",
+			input: []byte{
+				0x00, 0x02, // AFI
+				0x01,                   // SAFI
+				0x2c, 0x26, 0x20, 0x01, // Prefix
 			},
 			wantFail: true,
 		},

--- a/protocols/bgp/packet/path_attributes_test.go
+++ b/protocols/bgp/packet/path_attributes_test.go
@@ -934,6 +934,40 @@ func TestDecodeMultiProtocolReachNLRI(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "MP_REACH_NLRI with invalid length",
+			input: []byte{
+				0x00, 0x02, // AFI
+			},
+			wantFail: true,
+		},
+		{
+			name: "MP_REACH_NLRI with invalid length 2",
+			input: []byte{
+				0x00, 0x02, // AFI
+				0x01,                                           // SAFI
+				0x10, 0x20, 0x01, 0x06, 0x78, 0x01, 0xe0, 0x00, // incomplete NextHop
+			},
+			wantFail: true,
+		},
+		{
+			name: "MP_REACH_NLRI without prefixes",
+			input: []byte{
+				0x00, 0x02, // AFI
+				0x01,                                                                                                 // SAFI
+				0x10, 0x20, 0x01, 0x06, 0x78, 0x01, 0xe0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, // NextHop
+				0x00, // RESERVED
+			},
+			expected: &PathAttribute{
+				Length: 21,
+				Value: MultiProtocolReachNLRI{
+					AFI:      IPv6AFI,
+					SAFI:     UnicastSAFI,
+					NextHop:  bnet.IPv6FromBlocks(0x2001, 0x678, 0x1e0, 0, 0, 0, 0, 0x2),
+					Prefixes: []bnet.Prefix{},
+				},
+			},
+		},
 	}
 
 	t.Parallel()
@@ -997,6 +1031,13 @@ func TestDecodeMultiProtocolUnreachNLRI(t *testing.T) {
 					},
 				},
 			},
+		},
+		{
+			name: "MP_UNREACH_NLRI with invalid length",
+			input: []byte{
+				0x00, 0x02, // AFI
+			},
+			wantFail: true,
 		},
 	}
 

--- a/protocols/bgp/packet/path_attributes_test.go
+++ b/protocols/bgp/packet/path_attributes_test.go
@@ -1051,7 +1051,7 @@ func TestDecodeMultiProtocolUnreachNLRI(t *testing.T) {
 			wantFail: true,
 		},
 		{
-			name: "MP_REACH_NLRI with invalid prefixes",
+			name: "MP_UNREACH_NLRI with invalid prefixes",
 			input: []byte{
 				0x00, 0x02, // AFI
 				0x01,                   // SAFI

